[MRG] Bug fix and new feature: fix implementation of average precision score and add eleven-point interpolated option

ndingwall opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

Fixes [#4577](https://github.com/scikit-learn/scikit-learn/issues/4577) and [#6377](https://github.com/scikit-learn/scikit-learn/issues/6377)
#### What does this implement/fix? Explain your changes.

This adds an optional `interpolation` parameter to both `auc` and `average_precision_score`. By default, the value is set to `'linear'` which replicates the existing behavior, but there is also a `'step'` option that implements the strategy described in [Stanford's Introduction to Information Retrieval](http://nlp.stanford.edu/IR-book/html/htmledition/evaluation-of-ranked-retrieval-results-1.html), and as discussed in [this blog post](https://github.com/roaminsight/roamresearch/blob/master/BlogPosts/Average_precision/Average_precision_post.ipynb). To generalize this, `auc` also offers a parameter to control the direction of the step-wise interpolation; I'm sure someone will find this useful at some point!
#### Any other comments?

At some point, I'd support switching the standard behavior to step-wise interpolation, but that will make the existing docstring tests fail. I don't want to change them without the support of the community!

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/scikit-learn/scikit-learn/7356)

<!-- Reviewable:end -->

ndingwall committed:
Adds support for step-wise interpolation to auc and average_precision_score

ndingwall committed:
Adds support for step-wise interpolation to auc and average_precision_score

ndingwall committed:
Fixed line length

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -149,15 +197,30 @@ def average_precision_score(y_true, y_score, average="macro",
+    interpolation : string ['linear' (default), 'step']
+        Determines the kind of interpolation used when computed AUC. If there
+        are many repeated scores, 'step' is recommended to avoid under- or
+        over-estimating the AUC. See `Roam Analytics blog post
```
This doesn't seem a good source to me. I'm happy with citing the IR book, though.

ndingwall replied:
No problem - I'll remove it and add the IR book to the references.

ndingwall replied:
@amueller Related question: in the references section, is it okay to include the link to my blog post? I'm never sure whether the references are to justify the implementation, or to provide useful guidance for users. I feel like my post might be helpful for the latter but not for the former!

amueller commented:
You can also cite the Pascal VOC paper. Can you please also add a reference to the paper I cited earlier about the relation of ROC curve and PR curve?

amueller commented:
Wait, so currently we are using linear interpolation? That is actually... wrong... see the paper I cited.
However, this looks like it's not the same as what the Pascal VOC computes, which cheats even more than we do.... hum...

amueller commented:
Could you please do a comparison of the VOC measure, the measure you implemented, the measure in the IR book and our current measure?
This is actually a different issue for PR and ROC curve, as the paper that I mentioned explains. For ROC curve, linear interpolation is fine, for PR curve, it is not - which is what the blog post complains about.
You could do stepwise interpolation on the PR curve, which is a pessimistic boundary, or you could do the real interpolation, as explained in the paper, but that requires a validation set.

Also, can you maybe add an example that compares the different methods. This is pretty subtle.
So for AUC, there is "linear interpolation", "step-wise interpolation" and "VOC smoothing" (aka cheating). There could also be the correct smoothing using a validation set, which would be good.

For PR, there is "linear interpolation" (which we currently do and is wrong), "VOC smoothing" which is even worse, "step-wise interpolation" which is a lower bound and the correct interpolation as explained in the paper, which requires a validation set.

amueller commented:
The IR book is also cheating (using the VOC method), the blog post is not cheating as far as I can see.

amueller commented:
Oh the blog post is by you? Then never mind my comment there ;)

So interestingly AP as defined on wikipedia is again different from this. mean AP is the mean over recall@k for all k. That means some of the sawtooth are skipped over, but not all.

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -55,6 +57,25 @@ def auc(x, y, reorder=False):
         If True, assume that the curve is ascending in the case of ties, as for
         an ROC curve. If the curve is non-ascending, the result will be wrong.
 
+    interpolation : string ['trapezoid' (default), 'step']
```
linear, not trapezoid ;)

ndingwall replied:
Good catch!

ndingwall committed:
Fix docstring from 'trapezoid' to 'linear'

amueller commented:
Ok so after thinking about it for another hour: you are implementing the wikipedia definition. We should not allow "linear" interpolation for ``average_precision`` as it is wrong. We should also allow "interpolated" average precision as many people use it.

So I'd add a parameter to ``average_precision`` that is ``interpolation="none""  and add the option "pascal_voc", and maybe also the "ir_book" only we should give them better names.

I wouldn't worry about AUC for now.

ndingwall commented:
Before I get on this flight, I want to clarify your last comment so I can spend some flight time working on it! I haven't read the paper yet, so apologies if some of this would be pre-empted by that.

##### Interpolation options

0. **`None`** (default) Just return the mean of the precision values (and ignore recall). 
0. [**`pascal_voc`**](http://homepages.inf.ed.ac.uk/ckiw/postscript/ijcv_voc09.pdf) (Pg 11) This appears to be the same as the 11-point interpolated average precision (11PIAP from here on) mentioned in the [IR book](http://nlp.stanford.edu/IR-book/html/htmledition/evaluation-of-ranked-retrieval-results-1.html). For each of 11 recall values *r*, we choose the precision that corresponds to the smallest recall greater than or equal to *r*. Then we return the mean of these precisions. 
0. **`ir_book`** I don't think either of us like the approach that excludes operating points based on comparing their precision to the precision of the subsequent operating point. I'm therefore not keen to implement something that I wouldn't encourage people to use, especially because it's going to be fiddly!

##### Comments: 

'None' falls into the same pitfalls as linear interpolation. If we only have two operating points, the arithmetic mean of the precision values is the midpoint of the straight line joining them. That's what I was trying to escape!

This problem doesn't apply to the 11PIAP because the recall values are specified, so although we're taking an arithmetic mean, we're at least comparing oranges and oranges.

But 11PIAP does introduce a different problem. Consider a model that assigns scores which permit recall values of 0.09, 0.19, 0.29, etc. This model is likely to appear to be much worse than one whose recall values are 0.11, 0.21, 0.31, etc. even if the first one has much better precision at those operating points. We could imagine a situation where including examples _at random_ would actually _improve_ the 11PIAP by bringing the recall above the arbitrary thresholds.

The step interpolation I proposed is really just an extension of the 11PIAP to arbitrarily many points. It doesn't introduce the bias that linear interpolation does, and it allows each operating point to fully contribute whatever its recall value is. So I'd still advocate for including that as an option!

ndingwall committed:
Remove reference to blog

amueller commented:
By "none" I meant what you did, i.e. step-wise interpolation to the left. it's not really interpolation, though.
I'm not sure what you mean by the "ignoring recall". I meant the wikipedia definition of average precision. That's what you implemented, and that's what I think the default should be.

I hope I'll have some plots later today to illustrate.

ndingwall commented:
Oh, that makes sense. I call it 'interpolated' because I'm thinking of it from a graphical perspective, but I see that the wiki article comes to the same calculation without explicitly considering the curve.

I'll recode it so that the wikipedia definition ($\operatorname{AveP} = \sum_{k=1}^n P(k) \Delta r(k)$) is the default and I'll add an option to use `interpolation='eleven_point'` interpolation instead (and point to the IR book which has a good section on that).

Presumably I'll need to change some tests since we'll be changing the default behavior. Is there a best-practice way to do that?

amueller commented:
check which tests fail, fix them. add a test that checks that the current behavior is correct, not like the previous one, possibly using one of the reported bugs.
then add a section to ``bugfixes`` in whatsnew that points out the change in the behavior.
Usually we don't make backward incompatible changes, but I would consider this a bugfix.
I would really love to also see a visual comparison of the different methods as an example (with how to do it and how not to do it). Not sure I find the time to do that.
Also see #7372.
Currently the plots of ``precision_recall_curve`` use ``plt.plot`` which basically plots linear interpolation, which is a bit odd.

amueller commented:
please also check #4936 to see if you can reuse any of that.

amueller commented:
sorry accidental close.

ndingwall committed:
Change interpolation options to None and 'eleven_point'

ndingwall committed:
Revert to original auc() function

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -197,18 +141,20 @@ def average_precision_score(y_true, y_score, average="macro",
-            Uses a step function to interpolate between operating points.
+    interpolation : None (default), or string ['eleven_point']
+
+        ``None``:
```
Can you check how this renders / whether sphinx spits out warnings?

ndingwall replied:
Seems to be okay.
<img width="761" alt="screen shot 2016-09-20 at 3 51 43 pm" src="https://cloud.githubusercontent.com/assets/7325675/18675412/3133f5bc-7f4a-11e6-9830-29039b0b1bc1.png">

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -197,18 +141,20 @@ def average_precision_score(y_true, y_score, average="macro",
-            Linearly interpolates between operating points.
-        ``'step'``:
-            Uses a step function to interpolate between operating points.
+    interpolation : None (default), or string ['eleven_point']
```
why a list?

ndingwall replied:
I was just copying the style from other places that includes acceptable strings in a list. But they had multiple options so I guess that made more sense. Should it read `interpolation : None (default), or string 'eleven_point'` or `interpolation : None (default), or string ['eleven_point', ]`?

amueller replied:
the first, I think.

ndingwall replied:
Done. See the screenshot below.

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -149,14 +141,32 @@ def average_precision_score(y_true, y_score, average="macro",
     .. [1] `Wikipedia entry for the Average precision
-           <https://en.wikipedia.org/wiki/Average_precision>`_
+           <http://en.wikipedia.org/wiki/Average_precision>`_
+    .. [2] `Stanford Information Retrieval book
```
Maybe add the Pascal VOC paper?

ndingwall replied:
This one? Page 11 describes the metric. http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.157.5766&rep=rep1&type=pdf

amueller replied:
yes

ndingwall replied:
Done.

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -149,14 +141,32 @@ def average_precision_score(y_true, y_score, average="macro",
+        ``'eleven_point'``:
+            For each of the recall values, r in {0, 0.1, 0.2, ..., 1.0},
+            compute the arithmetic mean of the first precision value with a
+            corresponding recall ≥ r. See the `Stanford Information Retrieval
```
the ``>=`` symbol is probably not a good idea here (you could declare an encoding for the file but I think just writing >= is better).

ndingwall replied:
Done.

amueller reviewed (changes requested):
Please add a non-regression test. Ideally it would be great if you could have a test against the output of the Pascal VOC code.

ndingwall committed:
Change >= symbol to '>=' string

ndingwall commented:
@amueller is there a python implementation of Pascal VOC? The Matlab implementation is in [this comment](https://github.com/scikit-learn/scikit-learn/issues/4577#issuecomment-92002471), but I can't find a Python version. I'd write one, but it's likely to be identical to what the implementation here so that's probably not helpful for testing purposes!

amueller commented:
There's no python version. Would you mind running the matlab version? There could be subtle differences.
Running it with octave should be fine. If that doesn't work / is too much work, we might find someone who has matlab, or not bother.

ndingwall commented:
I don't have either matlab or octave. If you/someone else can run it and paste the `y_true` and `y_score` they used into this thread, I'd be happy to add a test.

ndingwall commented:
There are a couple of problems in the `test_ranking.py` file. 

0. [`_test_precision_recall_curve`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/tests/test_ranking.py#L473) uses `auc` on the precision & recall values as the baseline for `average_precision_score`. That means my implementation fails the test. Is there a sensible alternative to `auc` that I should compare against? Perhaps a slow implementation that more obviously follows the wiki article:

	```
	precision, recall, threshold = precision_recall_curve(y_true, y_score)
	precision = list(reversed(precision))
	recall = list(reversed(recall))
	average_precision = 0
	for i in range(1, len(precision)):
	    average_precision += precision[i] * (recall[i] - recall[i - 1])
	```

0. The [alternative implementation of average precision](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/tests/test_ranking.py#L101) suffers from a similar problem to the original `average_precision_score` implementation. In the edge case of identical predictions, the ordering of the documents matters.
	
	For example: 
	
	```
	_average_precision(np.array([1, 0, 0, 0]), 
	                   np.ones(4))
	```
	
	results in an average_precision of 0.25, while
	
	```
	_average_precision(np.array([0,0,0,1]), 
	                   np.ones(4))
	```
	
	results in 1.0. 
	
	This doesn't stop it passing the test, since it works whenever the elements `y_score` are mostly unique, and that implementation is only used to test against `auc`, and not  `average_precision_score`. Which is kind of weird in itself! But this raises a question: should the test implementation approximate the actual implementation for all inputs? If that's the case, then we should replace it with the (inefficient) implementation above. Or should it just work for a specified test case? In which case I'll leave it as it is.

amueller commented:
1. slow and obvious implementation sounds good.
2. If the test currently works, I'd say leave it and comment that the alternative implementation only works in this edge-case.

amueller commented:
@ogrisel @jnothman @GaelVaroquaux I would kinda like this in the final version as our current average precision is not great :-/

@ndingwall would you be interested in adding a plot of a PR curve comparing the 11-point interpolation, the IR "skip dips", the "linear interpolation" we had before and the "left step interpolation" you're implementing here?
I'd rather get this merged sooner and maybe backport to 0.18 than waiting for an example, though if you have extra time on hand, I'd love to see the plot in the documentation.

ndingwall commented:
@amueller I might be able to find some time over the weekend to amend the example. I started it (see #7372) so perhaps we should discuss it there? I'm reluctant to include the "skip dips" part because that's going to require a load of new code and I'm yet to see a justification for what seems to me (and to you as well I think) as cheating.

ndingwall committed:
Small change to docstring

ndingwall committed:
Add reference to Pascal VOC book

ndingwall committed:
Replace auc with slow implementation of average precision for testing

amueller commented:
cool let's see what the CIs say. I should do something else now, though and might not get back to you on the matlab thing soon (the final final version of my book is due on monday and I have a keynote next week)

ndingwall committed:
Minor change to docstring

ndingwall committed:
Update example to include 'eleven_point' interpolation

ndingwall committed:
Update example to include 'eleven_point' interpolation

ndingwall commented:
The most recent (annoyingly duplicated) commit is related to #7372

ndingwall committed:
Added a test for interpolated average precision

amueller commented:
thanks for your updates. Travis is unhappy. I'll review next week (I'm travelling)

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -172,16 +184,57 @@ def average_precision_score(y_true, y_score, average="macro",
         precision, recall, thresholds = precision_recall_curve(
             y_true, y_score, sample_weight=sample_weight)
-        return auc(recall, precision)
+        return -sum(np.diff(recall) * np.array(precision)[:-1])
```
Please use np.sum

ndingwall replied:
Done.

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -149,14 +141,34 @@ def average_precision_score(y_true, y_score, average="macro",
     sample_weight : array-like of shape = [n_samples], optional
         Sample weights.
 
+    interpolation : None (default), or string 'eleven_point'
```
better without `string`

ndingwall replied:
Done. Although since we only have one option for `interpolation`, would this be better as a bool? So `interpolation=False` is the default and corresponds to the `None` option here, and `interpolation=True` corresponds to `interpolation='eleven_point'`. We could even change the parameter name to `eleven_point_interpolate` to make it extra clear. I have no particular preference, except that I could imagine people having to repeatedly check the docstring to remember that they should be typing `'eleven_point'`.

amueller replied:
I think having strings is better because we might add others in the future, in particular the "correct" way ;)

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -172,16 +184,57 @@ def average_precision_score(y_true, y_score, average="macro",
+        eleven_precisions = list()
+        # Don't use the first (precision, recall) pair with precision = 1.
+        start = 1
+        for target_recall in np.arange(0, 1.1, 0.1):
```
Can't you do this with something along the lines of `precision[numpy.searchsorted(recall, np.arange(0, 1.1, 0.1))]`?

ndingwall replied:
Oh that's neat - I didn't know about `searchsorted` before.

jnothman replied:
Once you start hacking numpy a bit, you realise it comes up all over the place. But Python's `bisect` is the same (just not as efficient for numeric arrays)

ndingwall committed:
Change to docstring

ndingwall committed:
Use np.sum instead of sum

ndingwall committed:
Use np.searchsorted instead of for loop

amueller commented:
CI is unhappy

ndingwall committed:
Remove redundant reference to auc 'interpolation' parameter

ndingwall committed:
Add lines around doc tests

ndingwall committed:
Change desired value for average_precision_score test

ndingwall committed:
Loads more changes to desired values in tests

ndingwall commented:
I've just added a load of changes to the desired values in `test_precision_recall_curve_toydata` which does a load of multi-label classification. I've only changed the `_binary_average_precision` function so I'm assuming that the changes in the scores are only because of the change to how this binary average precision is calculated, but someone might want to double check that.

ndingwall committed:
One more change to a test's desired value

ndingwall committed:
Change docstring to match new average_precision_score implementation

ndingwall committed:
Use model_selection instead of cross_validation

ndingwall commented:
@amueller I'm not sure why ci is failing this time. It worked on previous commits. The log talks about a server hangup when it's running `plot_stock_market.py` which isn't affected by the changes in this PR. So I suspect it just needs to be re-run, but I don't know how to initiate that without submitting another commit.

amueller commented:
you can always amend the last commit and force push.

ndingwall commented:
@amueller Done - all checks are passing now. Anything else I need to do?

amueller commented:
no, "just" needs reviews now.

amueller commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -494,10 +534,15 @@ def test_precision_recall_curve_pos_label():
+    assert_equal(interpolated_average_precision,
+                 average_precision_score(y_true, probas_pred,
+                                         interpolation='eleven_point'))
     assert_almost_equal(_average_precision(y_true, probas_pred),
```
why are the precisions in the "assert_almost_equal" so low? Can you try to remove them and see if it still works?

ndingwall replied:
Those were the values that were there before. I'll try and see what happens.

ndingwall replied:
One of the tests is failing on circleci. When I run it on my machine, I get an average precision of 0.85873... but circleci reports 0.85899.... Perhaps that's why the precisions were set low.

ndingwall replied:
The other checks are still failing with a lower tolerance on `assert_almost_equal`. For some reason, `average_precision_score` is coming out as 0.85899092682932521 in travis. On my machine, I get 0.85873118656958514, which is equal to 12 decimal places with the 'slow' implementation in `test_rankings.py`. I can't recreate the 0.85899... number. Any ideas? The difference seems too big to be a 32 vs 64 bit problem, but we are adding up lots of decimals (that are themselves products of decimals...)

amueller commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -123,6 +129,40 @@ def _average_precision(y_true, y_score):
+    precision = list(reversed(precision))
+    recall = list(reversed(recall))
+    eleven_precisions = list()
+    for threshold in np.arange(0, 1.1, 0.1):
```
why not searchsorted?

ndingwall replied:
I'll leave this change until we work out how we want to implement 11-point average precision.

amueller reviewed (changes requested):
Looks pretty good apart from some possible simplifications

amueller commented:
@ndingwall I would really appreciate if you had time to polish this. Sorry for the long turn-around.

ndingwall commented:
Sorry for the slow reply, @amueller - I was at NIPS when your last batch of comments came through and it's taken a while to catch up on things! I should have some time over the weekend or early next week to work on this so I'll push some changes soon!

amueller commented:
That's a pretty good excuse, I'd say ;) No worries and looking forward to your changes :)

jnothman commented:
IIUC this PR no longer intends to make changes to AUC, but acts to:
* fix a bug average precision score where its calculation was invalid/non-standard. (Are we sure its linear interpolation is never appropriate? I've not thought about it.)
* add an additional mode of interpolation

In the first case, please update the title and description of the PR.

Additionally, model_evaluation.rst describes each of the metrics available, and should be updated more extensively to describe the interpolation issue and the differences between the approaches.

ndingwall committed:
Improved description of average precision and interpolated average precision

ndingwall commented:
Thanks @jnothman. I have a few other updates to make (I haven't forgotten, @amueller!) so I'll make those changes at the same time.

ndingwall committed:
Use searchsorted to get circle coordinates

ndingwall committed:
Use plt.fill_between instead of custom function

ndingwall commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -111,44 +114,89 @@
+interpolated_average_precision = dict()
 for i in range(n_classes):
-    precision[i], recall[i], _ = precision_recall_curve(y_test[:, i],
+    precision[i], recall[i], _ = reversed_precision_recall_curve(y_test[:, i],
```
I can't remember what the original reason for it was: probably something to do with having recall in ascending order. I've modified `get_circle_coords` to use `searchsorted` as you suggested but this requires that recall be in ascending order anyway, so I need to reverse it at some point. Where would be the best place to do this?

GaelVaroquaux replied:
I don't understand the need for reversing. It's a pity, because it makes the example harder to understand. That example is already getting a bit challenging: not an ideal introduction to precision-recall.

ndingwall commented:
Sorry for the delay in coming back to this, @amueller! I'm kind of glad it's been a while though  because I've noticed a mistake in my implementation of eleven point average precision. I use the _first_ operating point above a given recall target, but the definition requires the _best_ operating point above a given recall target (as determined by the highest precision).

Here's the relevant part of [the Pascal VOC rubric](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.157.5766&rep=rep1&type=pdf):

<img width="576" alt="screen shot 2017-01-29 at 7 32 45 pm" src="https://cloud.githubusercontent.com/assets/7325675/22406833/7083bb68-e65a-11e6-8a90-d3565d0f84bf.png">

But this seems to be ill-defined in the r = 0 case: we get 0 recall by assigning everything to the negative class. The precision is the fraction of positively-classified documents that are actually positive, or TP / (TP + FP). We're assuming that recall = 0, so TP must be 0 too (since recall = TP / (TP + FN)). So precision should be either 0 or undefined.

But the table in the [IR book](http://nlp.stanford.edu/IR-book/html/htmledition/evaluation-of-ranked-retrieval-results-1.html) assigns a precision of 1 when recall is 0. The Pascal VOC document references _Introduction to Modern Information Retrieval_ by G. Salton, M. J. Mcgill (1986) but I can't find it online. Perhaps that would shed some light on this...

Either way it seems like we'll need to implement a `p_interp` function which would just be the max of the sub-list of _p_ corresponding to _r_ >= _r*_ for each _r*_ in [0, 0.1, 0.2, ... 1.0].

ndingwall commented:
@jnothman I realized I didn't respond to your first question:

> Are we sure its linear interpolation is never appropriate? I've not thought about it.

Yes. I wrote a [blog post](http://roamanalytics.com/2016/09/07/stepping-away-from-linear-interpolation/) about this and that's what sparked this PR. (Some images aren't appearing but they're not relevant - if you want to read it you can safely start at the _Average precision_ section). But the basic conclusion is that interpolation is equivalent to arbitrary tie-breaking, and random tie-breaking leads to horizontal lines. (If we can do something better than arbitrarily tie-break, then we'd just have more operating points to plot and now we'd need to interpolate between them - by arbitrarily tie breaking).

ndingwall commented on sklearn/metrics/ranking.py:
```diff
@@ -110,10 +102,12 @@ def auc(x, y, reorder=False):
 
-    This score corresponds to the area under the precision-recall curve.
+    Optionally, this will compute an eleven-point interpolated average
+    precision score, which fixes recall values and averages precisions
```
Yeah, I find it a bit hard to explain in a sentence! How about

> Optionally, this will compute an eleven-point interpolated average precision score: for 11 target recall values ([0, 0.1, 0.2, ... 1.0]), this averages the best precision values possible while meeting or exceeding the target recall.

GaelVaroquaux replied:
This is better. +1 for such as change.

ndingwall committed:
Revert change to y_score specification in docstring

ndingwall commented on sklearn/metrics/ranking.py:
```diff
@@ -127,8 +121,7 @@ def average_precision_score(y_true, y_score, average="macro",
         Target scores, can either be probability estimates of the positive
-        class, confidence values, or non-thresholded measure of decisions
-        (as returned by "decision_function" on some classifiers).
+        class, confidence values, or binary decisions.
```
No idea! I'll revert it.

ndingwall committed:
Docstring improvements

ndingwall commented on sklearn/metrics/ranking.py:
```diff
@@ -172,16 +185,49 @@ def average_precision_score(y_true, y_score, average="macro",
+
+    def _binary_eleven_point_average_precision(
+            y_true, y_score, sample_weight=None):
+        precision, recall, thresholds = precision_recall_curve(
```
I hadn't noticed that... And no, I guess there's no test (since they all passed!)

ndingwall committed:
Removed unused variables

ndingwall committed:
Use sample_weight for eleven point average precision

ndingwall committed:
Use correct p_interp function to compute 11-point AP

ndingwall committed:
Update example to use p_interp in 11-point AP

ndingwall committed:
Update test for interpolated average precision to use p_interp

ndingwall committed:
Bug fix in test

ndingwall committed:
Make tests more precise

ndingwall committed:
Comment to explain that calculation is step function integral

ndingwall commented on sklearn/metrics/ranking.py:
```diff
@@ -39,26 +39,20 @@
 
 def auc(x, y, reorder=False):
     """Compute Area Under the Curve (AUC) using the trapezoidal rule
-
```
That's my IDE trying to be clever when I copied/pasted. I've restored them.

ndingwall committed:
Restore newlines in auc docstring

ndingwall committed:
Change precision of test

ndingwall commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -494,19 +531,24 @@ def test_precision_recall_curve_pos_label():
     # Smoke test in the case of proba having only one value
     p, r, thresholds = precision_recall_curve(y_true,
                                               np.zeros_like(probas_pred))
     precision_recall_auc = auc(r, p)
```
Incidentally, this test (part of `_test_precision_recall_curve`) uses the old, incorrect interpretation of average precision. But it's not actually claiming to be average precision, so should we leave it as it is?

amueller commented:
I'm basically not doing any sklearn stuff this month and maybe next month because.. time... but this is on top of my priority list... So are they actually cheating by picking the best points?

GaelVaroquaux commented on sklearn/metrics/ranking.py:
```diff
@@ -172,16 +190,48 @@ def average_precision_score(y_true, y_score, average="macro",
+        recall = recall[-2::-1]
+
+        return np.mean([precision[i:].max() for i in
+                        np.searchsorted(recall, np.arange(0, 1.1, 0.1))])
```
It would be good to move those two functions outside the enclosing function. In general, functions defined inside functions is not great style, as it hinders reuse and testing.

varunagrawal replied:
@ndingwall I believe having them at the module level or at `metrics.base` would be preferred.

GaelVaroquaux commented:
I've made a bunch of minor comments. Overall, I think that this PR is good and that it is very important to merge it (one of my students just hit the corresponding bug :D ).

The only aspect that I feel uneasy about is the example: it was just made much more complicated.

There is also one failing test, due to numerical instablity. It should be corrected.

Finally (and that's quite important), a clear note should be made in the docs/whats_new.rst file to stress the change and explain it. Some users will be very surprised when they update scikit-learn and find that they get different scores.

@ndingwall: do you think that you can address those comments? And I'll merge your PR. Thanks a lot!

ndingwall commented:
@GaelVaroquaux thanks for the comments! I should be able to get to them this week.

I agree about the example - perhaps we could split it into two: one for the binary case and one for multi-class? I feel like the current one is complicated partly because it's trying to cram so much in! On the other hand, having multiple examples for a single function is probably confusing in itself. I can't remember the justification for reversing of precision and recall lists, but I'll try and rework it without that either way.

ndingwall commented:
@amueller I realize I didn't respond to your question before:

> So are they actually cheating by picking the best points?

That sums up how I feel about, but there may be some subtlety to the justification that I haven't grasped. In the [IR book](https://nlp.stanford.edu/IR-book/html/htmledition/evaluation-of-ranked-retrieval-results-1.html) they say 

> The justification is that almost anyone would be prepared to look at a few more documents if it would increase the percentage of the viewed set that were relevant (that is, if the precision of the larger set is higher).

but (as I think we discussed before), you'd only be prepared to look at them if you already knew they were relevant. If your classifier isn't giving you that information (and it wouldn't, since it must have assigned them lower relevance than the things it's already shown you), it doesn't seem reasonable to use that information in evaluating the classifier.

GaelVaroquaux commented:
> I agree about the example - perhaps we could split it into two: one for
> the binary case and one for multi-class? I feel like the current one is
> complicated partly because it's trying to cram so much in! On the other
> hand, having multiple examples for a single function is probably
> confusing in itself.

I think that one good way to go about this is to split the example in
two, but in the same file, using the restructure text formating, as in
http://scikit-learn.org/stable/auto_examples/applications/plot_out_of_core_classification.html

ndingwall committed:
Better description of 11-point average precision

ndingwall commented on sklearn/metrics/ranking.py:
```diff
@@ -172,16 +190,48 @@ def average_precision_score(y_true, y_score, average="macro",
+        recall = recall[-2::-1]
+
+        return np.mean([precision[i:].max() for i in
+                        np.searchsorted(recall, np.arange(0, 1.1, 0.1))])
```
I put the second function there to reflect the existing code which has `_binary_average_precision` inside the `average_precision_score` function. I'm happy to move them out, but they're both very specific to average precision so I doubt they'd ever be reused! Would you prefer them at the module level (prefixed with `_`) anyway?

ndingwall committed:
Separate eleven-point interpolated average precision in example

ndingwall committed:
Fix typo

ndingwall committed:
Add changes to average_precision_score

ndingwall commented:
@GaelVaroquaux, I finally got to addressing your comments! 

I ran `nosetests` and everything seems to be passing. I get a load of warnings about `metrics/classification.py` and a couple of `DeprecationWarning`s but the average precision tests are passing for me. Is the failing test the one I refer to [in this comment](https://github.com/scikit-learn/scikit-learn/pull/7356/files/1b4539c4e591270fd458772f25450445a02fb063#r98456142)?

I wasn't sure where to add the changes in `whats_new.rst` so I created a `0.18.2` section. I can move them if this will be merged into a different version.

I followed your advice to split the example into sections, but I decided to move eleven point average precision entirely into its own section. This both avoids confusing the original average precision computation, and provides a much clearer explanation of eleven point AP. I printed out a string table which seems a bit ugly but I think helps to explain the algorithm. Let me know if I should remove it or change it to something else.

amueller commented:
@ndingwall I'm back! test are failing.

For testing the 11-point average, I would probably like to test against actual outcomes of the Pascal VOC code, since that is de-facto the definition of this measure (because defining scientific evaluation measures with matlab code is such a good idea!).

amueller commented on doc/whats_new.rst:
```diff
@@ -6,6 +6,34 @@
 Release history
 ===============
 
+Version 0.18.2
```
This version will not happen, we'll do 0.19 soonish?

ndingwall commented:
@amueller The CircleCI output seems to stop with this error so I'm not seeing outputs of tests, but I had [a problem before](https://github.com/scikit-learn/scikit-learn/pull/7356/files/1b4539c4e591270fd458772f25450445a02fb063#r98456142) that my local results don't match what CircleCI is reporting.

```
Exception occurred:
  File "/home/ubuntu/scikit-learn/doc/sphinxext/sphinx_gallery/docs_resolv.py", line 48, in _get_data
    with open(url, 'r') as fid:
IOError: [Errno 2] No such file or directory: '/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/generated/sklearn.feature_selection.SelectKBest.rst.html'
```

> For testing the 11-point average, I would probably like to test against actual outcomes of the Pascal VOC code, since that is de-facto the definition of this measure

Do you have an example? I don't have Matlab so can't generate any.

jnothman commented:
> I wasn't sure where to add the changes in whats_new.rst so I created a 0.18.2 section. I can move them if this will be merged into a different version.

If you merge with current master, you'll not only have an opportunity to fix current conflicts, but find that there is a section for 0.19

jnothman commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -54,11 +54,32 @@
+/viewdoc/download?doi=10.1.1.157.5766&rep=rep1&type=pdf>`_. In the example
+below, the eleven precision values are indicated with an arrow to pointing to
+the best precision possible while meeting or exceeding the desired recall.
+Note that it's possible that the same operating point might correspond to
```
This is in contradiction to the definition of operating point above. Do you simply mean that precision may be identical for multiple recall values?

ndingwall replied:
No, it's that there's no requirement that the eleven operating points chosen must be distinct: e.g. the points corresponding to recalls of 0.2, 0.3 and 0.4 might all be the same operating point at e.g. (0.45, 0.6) if 0.6 is the best precision you can get with a recall of at least 0.2.

jnothman commented on doc/whats_new.rst:
```diff
@@ -6,6 +6,34 @@
+   - :func:`metrics.ranking.average_precision_score` no longer linearly
+     interpolates between operating points, and instead weights precisions
+     by the change in recall since the last operating point, as per the
+     `Wikipedia entry <http://en.wikipedia.org/wiki/Average_precision>`_.
```
I think you need to cite a more stable reference; at a minimum, a versioned Wikipedia page.

ndingwall replied:
👍🏻

jnothman reviewed (commented):
I've not reviewed it all, but I don't see how we can make a fundamental change to basically *every* evaluation of a metric without having a deprecation cycle or something even more drastic like giving the new implementations new function names. People have, no doubt, reported experimental results with the incumbent implementation and will find their measures suddenly and inexplicably changed on update. I don't think that's okay.

I'm pretty sure I recall an earlier version of this maintaining current behaviour, and I think we need to keep that, with a warning, for some time; or else warn loudly that it has changed but instruct the user how to silence the warning.

amueller commented:
Can you maybe fix the conflicts?

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -172,16 +192,48 @@ def average_precision_score(y_true, y_score, average="macro",
-    0.79...
+    0.83...
+
+    >>> yt = np.array([0, 0, 1, 1])
```
you could also just use the y_true and y_score variables defined above, instead of redefining them here?

amueller reviewed (commented):
Looks good apart from the minor nitpick, I think. Trying to get @agramfort to dig up his matlab.

amueller commented:
I didn't realize Pascal VOC uses the standard AP, not 11 point. Is there a reference implementation for the 11 point somewhere?

agramfort commented:
I did:

```
import numpy as np
from sklearn.datasets import load_iris

iris = load_iris()
X, y = iris.data, iris.target
np.savetxt('X.txt', X[:, :1])
np.savetxt('y.txt', y)
```

then in matlab using VOC code:

```
load X.txt
load y.txt

y(y == 1) = -1;
y(y == 2) = 1;
out = X;
gt = y;

[so,si]=sort(-out);
tp=gt(si)>0;
fp=gt(si)<0;

fp=cumsum(fp);
tp=cumsum(tp);
rec=tp/sum(gt>0);
prec=tp./(fp+tp);

ap=VOCap(rec,prec)
```

returns

```
ap =

    0.7893
```

GaelVaroquaux commented:
In the interest of getting this merged while the heat of the sprint lasts, I am taking over this PR and finishing up the last details.

ndingwall commented:
@amueller The [Pascal VOC documentation](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.157.5766&rep=rep1&type=pdf) is very clear that they're using 11-point AP in their evaluation (top right of page 11). It's not dated so this might have changed - do you have the actual Matlab code used in their evaluation?

Will you or @GaelVaroquaux respond to @jnothman's [comment](https://github.com/scikit-learn/scikit-learn/pull/7356#pullrequestreview-40044129) above?

GaelVaroquaux commented:
I want to heavily rewrite the example. It's gotten way too complicated for demoing the average precision. I want to try to break it down more (and maybe do slightly less pretty plots).

amueller commented:
@ndingwall http://host.robots.ox.ac.uk/pascal/VOC/voc2012/VOCdevkit_18-May-2011.tar is the matlab code... hm

ndingwall commented:
@amueller, it looks like they're using _interpolated_ average precision, just not 11-point interpolated. This loop propagates the best precision backwards iteratively from the highest recall value.

```
for i=numel(mpre)-1:-1:1
    mpre(i)=max(mpre(i),mpre(i+1));
end
```

But either way, this contradicts the Pascal VOC documentation I've linked to in the docstrings! Do you know if this was updated somewhere?

ndingwall commented:
@GaelVaroquaux: agreed. It's hard to find the balance between providing examples of all functionality, and avoiding unnecessary complication. You're better placed than me to strike that balance! In any case, it looks like we might be able to abandon 11-point interpolated AP, although we might have to replace it with interpolated AP which suffers the same weird peeking-at-the-test-set-performance problem that 11-point IAP does.

GaelVaroquaux commented:
So, it seems that on different datasets, we do not reproduce well the pascal VOC measure. I am not sure why

ndingwall commented:
@GaelVaroquaux, that's probably because I based the code on [this documentation](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.157.5766&rep=rep1&type=pdf), which doesn't correspond to their evaluation code. This calls for 11-point interpolated AP, but the code does straight interpolated AP (no fixing of recall values). I could probably amend the code to reflect their _code_, but it'd be good to also update the link to the documentation. I just tried searching for an updated version of the documentation but the host.robots.ox.ac.uk website seems to be down at the moment...

GaelVaroquaux commented:
@ndingwall : do you mean that it should match our "interpolation=None"? That's not what I am observing :$

amueller commented:
no, interpolation=None is pessimistic, while I think theirs is optimistic (aka cheating)

ndingwall commented:
@GaelVaroquaux - as @amueller said, there's does use interpolation, just not the 11-point variety. We don't have an implementation that matches their evaluation code but it should be easy to write.

GaelVaroquaux commented:
@ndingwall : understood. I confirm looking at their code.

It doesn't seem to be useful to try to replicate it.

ndingwall commented:
@GaelVaroquaux, the interpolated version is requested in https://github.com/scikit-learn/scikit-learn/issues/4577. I don't mind removing it (and the 11-point version) because, like @amueller, I consider them both to be cheating! But we could argue that the code should reflect what the community uses rather than what we think is best...

GaelVaroquaux commented:
Closing in favor of #9091

amueller commented:
@GaelVaroquaux reopen for 11 point and/or general interpolation?

varunagrawal commented:
@GaelVaroquaux you have closed this issue and stated in #9091 that you won't be finishing that PR. Can we please re-open and resolve this issue?

jnothman commented:
@vene, it seems you and @gaelvaroquaux discussed 11-point average precision
at a sprint last year and thought it was problematic. Could you please
enlighten us?

varunagrawal commented on sklearn/metrics/ranking.py:
```diff
@@ -172,16 +192,48 @@ def average_precision_score(y_true, y_score, average="macro",
+        # We need the recall values in ascending order, and to ignore the first
+        # (precision, recall) pair with precision = 1.
+        precision = precision[-2::-1]
+        recall = recall[-2::-1]
```
I think it would make sense to fix line 470 in `precision_recall_curve` instead of doing this. Please refer to #8280

varunagrawal reviewed (commented):
Collation of related issues in order to better improve the PR.
